dali/pipeline/operator/builtin/merge.cc

```diff
@@ -0,0 +1,203 @@
+
+  for (int output_sample_idx = 0; output_sample_idx < predicate.num_samples();
+       output_sample_idx++) {
+    int input_group_idx = get_group_index(predicate, output_sample_idx);
+    auto &input = ws.template Input<Backend>(input_group_idx);
+
+    // get the index within input group and increment for the next occurrence.
+    int input_sample_idx = sample_idx_in_input[input_group_idx];
+    sample_idx_in_input[input_group_idx]++;
+
+    if (std::is_same_v<Backend, GPUBackend> || input.is_pinned() == *pinned_) {
+      output.SetSample(output_sample_idx, input, input_sample_idx);
+    } else {
+      // Pessimistic variant, we need to copy.
+      // TODO(klecki): Do one allocation, where samples that we share are 0-volumed - this might
+      // be perf optimization reducing the number of allocations to 1.
+      CopySampleToOutput(output, output_sample_idx, input, input_sample_idx, ws);
+    }
+  }
+  FinalizeCopy(ws);
+  WriteTestsDiagnostics(ws);
+}
+
+template <>
+void Merge<CPUBackend>::CopySampleToOutput(TensorList<CPUBackend> &output, int output_sample_idx,
+                                           const TensorList<CPUBackend> &input,
+                                           int input_sample_idx, Workspace &ws) {
+  auto &tp = ws.GetThreadPool();
+  tp.AddWork(
+      [&output, &input, output_sample_idx, input_sample_idx](int thread_idx) {
+        output.ResizeSample(output_sample_idx, input.shape()[input_sample_idx]);
+        output.CopySample(output_sample_idx, input, input_sample_idx, output.order());
+      },
+      volume(input.tensor_shape_span(input_sample_idx)));
+}
+
+
+template <>
+void Merge<CPUBackend>::FinalizeCopy(Workspace &ws) {
+  ws.GetThreadPool().RunAll();
+}
+
+
+template <>
+void Merge<GPUBackend>::CopySampleToOutput(TensorList<GPUBackend> &output, int output_sample_idx,
+                                           const TensorList<GPUBackend> &input,
+                                           int input_sample_idx, Workspace &ws) {
+  assert(false && "This codepath should not be executed");
+}
+
+
+template <>
+void Merge<GPUBackend>::FinalizeCopy(Workspace &ws) {
+  assert(false && "This codepath should not be executed");
+}
+
+
+template <typename Backend>
+void Merge<Backend>::RegisterTestsDiagnostics() {
+  this->RegisterDiagnostic("input_0_pinned", &in_0_pinned_);
+  this->RegisterDiagnostic("input_1_pinned", &in_1_pinned_);
+  this->RegisterDiagnostic("output_pinned", &out_pinned_);
+}
+
+
+template <typename Backend>
+void Merge<Backend>::WriteTestsDiagnostics(const Workspace &ws) {
+  in_0_pinned_ = ws.template Input<Backend>(0).is_pinned();
+  in_1_pinned_ = ws.template Input<Backend>(1).is_pinned();
+  out_pinned_ = ws.template Output<Backend>(0).is_pinned();
+}
+
+
+DALI_SCHEMA(_conditional__Merge)
+    .DocStr(R"code(Merge batch based on a predicate.)code")
+    .NumInput(2)
+    .NumOutput(1)
+    .AddArg("predicate", "Boolean categorization of the inputs", DALI_BOOL, true)
```

Comment:
There is no simple usage, as this is internal now. What do you mean if the predicate can be iterable or generator? It can support whatever DALI pipeline supports for argument input, provided it has correct type and shape.

Reply:
If we expose them at some point, we can write the doc based on how they should be used, but for now I think I need some more data about it.